Blog: How GitHub and GitLab use feature flags (#6384)
PostHog/posthog.com: contents/blog/github-gitlab-feature-flags.md

---

> title: Lessons from how GitHub and GitLab use feature flags

Reviewer: I like the suggestion "What you can learn from how GitHub and GitLab user features flags". As the reader, It suggests to me that there are actionable things to takeaway from the post.

> title: Lessons from how GitHub and GitLab use feature flags

Author: updated

date: 2023-07-26
author: ["ian-vanagas"]
showTitle: true
rootpage: /blog
sidebar: Blog
hideAnchor: true

> featuredImage: ../images/blog/green-blog-image.jpg

Author: https://github.com/PostHog/posthog.com/issues/6412 (or can I just do this myself 🤔 )

- Scripts to check for and remove flags. [GitLab](https://docs.gitlab.com/ee/development/feature_flags/controls.html/#cleanup-chatops) has a "chatops" removal command (`/chatops run feature delete some_feature`). [GitHub](https://github.blog/2021-04-27-ship-code-faster-safer-feature-flags/#the-cost-of-a-feature-flag) has a script that uses regular expressions and `git grep` to find flags and modify the code using `rubocop-ast`.
- GitLab scans to find old flags (6 months or older) and initiate processes for making them permanent or removing them.

[GitLab](https://about.gitlab.com/handbook/product-development-flow/feature-flag-lifecycle/#the-cost-of-feature-flags) specifically finds the cost of not having feature flags is higher than having them. When an issue happens with non-feature flagged code, they must revert the release, clean up the related code, and ship a fix to unblock future releases. Feature flags are a larger upfront cost, but are "cheaper" to rollback, saving time and energy. 

> ![Costs](../images/blog/github-gitlab-feature-flags/cost.png)

Author: Yeah, it was based on how GitLab explains the cost-benefit of feature flags, I wanted to add it to add some visuals to the post but it doesn't work super well. I'll just remove it.

## Takeaways from GitHub and GitLab's usage of feature flags

- Ensuring safe deployment and delivery of code is critical at scale.

- Feature flags support this goal by limiting the downside of issues in deployed code. This is done by disconnecting deployment from rollout and enabling simple rollbacks. 

- Feature flags have costs, but for changes that are high risk, in high traffic areas, introduce external services, or add new usage restrictions, those costs are worth the benefits. 

- In large organizations, feature flags work better for collaboration than branches.

- Users aren't the only possible "target" for feature flags. Organizations, teams, repositories, projects, and apps can all be targets too. This ensures a consistent experience.

- Rolling out a new feature with flags requires coordination and planning. Creating a process to help each stage of rollout lowers the risk of issues. This process might include documentation, change management, and notifying stakeholders.

> - Having commands and scripts for adding, enabling, disabling, and removing feature flags creates consistency and limits technical debt.

Author: Combined 1 and 2, added more to 4.

Although we covered feature flag [use cases](/blog/feature-flag-benefits-use-cases) and [best practices](/blog/feature-flag-best-practices), it is always useful to learn how large companies work with feature flags. Two leading examples are GitHub and GitLab. This post details why, when, and how they use feature flags.

## Why do GitHub and GitLab use feature flags?

Both GitHub and GitLab are massive software companies. They are both at the forefront of software development and DevOps, so maintaining best practices is important. Broadly, this means making the deployment of code safer and more consistent.

In GitHub’s case, they frame this as "**lowering deployment risk**." They want deployments to be risk-free. Putting new changes behind a flag lowers the risk of failing deployments. If there is an issue with a feature, you disable the flag rather than needing to roll back the entire deployment.

> In GitLab’s case, they have a "**progressive delivery mindset**." They find feature flags are a great tool for incremental, continuous delivery, allowing them to separate deployment and rollout. This is especially valuable to them because deployment is inconsistent due to their self-hostability. Feature flags increase the throughput of changes shipped as well as psychological safety for developers.

Reviewer: ```suggestion
In GitLab’s case, they have a "**progressive delivery mindset**." They find feature flags are a great tool for incremental, continuous delivery, allowing them to separate deployment and rollout. This is especially valuable to them because deployment is inconsistent due to their self-hostability. Feature flags increase the speed of shipping changes as well as provide psychological safety for developers.
```

## When do they use feature flags?

Because GitHub and GitLab share similar reasons for using feature flags, they also share similar scenarios when they use feature flags.

- **High risk:** High traffic areas or queries and significant UI, database, or storage changes.
- **Restrictions and externalities:** Security, permission restrictions, and third-party dependencies.
- **Working incrementally on new features:** Developers collaborate with flags instead of branches. This keeps changes small, makes reviewers easier, and prevents merge conflicts.

This means areas **that might not use flags** include brand-new features or APIs, low-traffic areas, and non-invasive frontend changes.

## What does their process for implementing feature flags look like?

> Both GitHub and GitLab have fairly standard use cases and implementations of the feature flags themselves. They call a service with a feature flag value and "actor," returns another value stating whether the flag is active or not. They use these as gates for code and features. 

Reviewer: ```suggestion
Both GitHub and GitLab have fairly standard use cases and implementations of the feature flags themselves. They call a service with a feature flag value and an "actor," which then returns a value stating whether the flag is active or not.
```

> **What is an "actor"?** Both GitLab and GitHub don’t exclusively target users with feature flags. They target "actors." These are users, organizations, teams, enterprises, repositories, projects, or apps. They use actors to create consistent experiences for a related group of users. For example, an organization should have a consistent experience for all its members.

A specific example of how GitHub uses feature flags is in the rollout of their [new GitHub API rate limiter](https://github.blog/2021-04-05-how-we-scaled-github-api-sharded-replicated-rate-limiter-redis/). This involved writing a completely new backend with Redis. They gated their new backend with a feature flag and rolled out slowly and smoothly. Once done, they removed the flag and old backend and integrated the new Redis backend further. Ironically, once they removed the flag, bug reports started rolling in causing them to have to rapidly fix a bunch of bugs (which is preventable with a feature flag). 

### GitLab’s feature flag rollout process

Although GitHub only details parts of its feature flag usage, GitLab details its entire process from idea to rollout to removal. 

> The process starts before writing any code. An engineer creates a rollout plan for the feature and flag. They [detail this in an issue](https://gitlab.com/gitlab-org/gitlab/-/blob/master/.gitlab/issue_templates/Feature%20Flag%20Roll%20Out.md). This issue contains owners, stakeholders, expectations, rollout steps for different environments (non-production, production, global), and rollback steps. They  refine the rollout plan with engineering and product managers.

Reviewer: ```suggestion
The process starts before writing any code. An engineer creates a rollout plan for the feature and flag. They [detail this in an issue](https://gitlab.com/gitlab-org/gitlab/-/blob/master/.gitlab/issue_templates/Feature%20Flag%20Roll%20Out.md). This issue contains owners, stakeholders, expectations, rollout steps for different environments (non-production, production, global), and rollback steps. They refine the rollout plan with engineers and product managers.
```
